File: Karan/Hash tables/Minimum Window Substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        from collections import Counter
        need=Counter(t)
        missing=len(t)
        left=start=end=0
        for right,c in enumerate(s,1):
            if need[c]>0:
                missing-=1
            need[c]-=1
            if missing==0:
                while left<right and need[s[left]]<0:
                    need[s[left]]+=1
                    left+=1
                if end==0 or right-left<end-start:
                    start,end=left,right
                need[s[left]]+=1
                missing+=1
                left+=1
        return s[start:end]

File: Karan/Hash tables/Minimum Window Substring.java
import java.util.*;
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() == 0 || t.length() == 0) return "";
        Map<Character, Integer> need = new HashMap<>();
        for (char c : t.toCharArray()) need.put(c, need.getOrDefault(c, 0) + 1);
        int required = need.size();
        int l = 0, r = 0, formed = 0;
        Map<Character, Integer> window = new HashMap<>();
        int[] ans = {-1, 0, 0};
        while (r < s.length()) {
            char c = s.charAt(r);
            window.put(c, window.getOrDefault(c, 0) + 1);
            if (need.containsKey(c) && window.get(c).intValue() == need.get(c).intValue()) formed++;
            while (l <= r && formed == required) {
                if (ans[0] == -1 || r - l + 1 < ans[0]) {
                    ans[0] = r - l + 1;
                    ans[1] = l;
                    ans[2] = r;
                }
                char ch = s.charAt(l);
                window.put(ch, window.get(ch) - 1);
                if (need.containsKey(ch) && window.get(ch).intValue() < need.get(ch).intValue()) formed--;
                l++;
            }
            r++;
        }
        return ans[0] == -1 ? "" : s.substring(ans[1], ans[2] + 1);
    }
}